codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/RubyRuntimePlugin.java

```diff
@@ -0,0 +1,181 @@
+    public interface WriteAdditionalFiles {
+        /**
+         * Called to write out additional files needed by this Plugin.
+         *
+         * @param context GenerationContext - allows access to file manifest and symbol providers
+         * @return List of the relative paths of files written, which will be required in client.rb.
+         */
+        List<String> writeAdditionalFiles(GenerationContext context);
+    }
+
+    /**
+     * Builder for {@link RubyRuntimePlugin}.
+     */
+    public static class Builder implements SmithyBuilder<RubyRuntimePlugin> {
+
+        private RenderAdd renderAdd;
+        private WriteAdditionalFiles writeAdditionalFiles =
+                (context) -> Collections.emptyList();
+
+        /**
+         * Used to copy a plugin ruby file into the generated SDK. The copied file
+         * must be a plugin class (implements call method) under the Plugins namespace.
+         * This method will apply the generated service's namespace to the plugin file.
+         *
+         * @param rubyFileName the file name (with path) of the ruby file to copy.
+         * @return Return the Builder
+         */
+        public Builder rubySource(String rubyFileName) {
+            this.writeAdditionalFiles = (context) -> {
+                try {
+                    Path path = Paths.get(rubyFileName);
+                    String relativeName = "plugins/" + path.getFileName();
+                    String fileName =
+                            context.settings().getGemName() + "/lib/"
+                                    + context.settings().getGemName()
+                                    + "/" + relativeName;
+                    String fileContent =
+                            new String(Files.readAllBytes(path), StandardCharsets.UTF_8);
+
+                    RubyCodeWriter writer = new RubyCodeWriter(context.settings().getModule());
+                    writer
+                            .openBlock("module $L", context.settings().getModule())
+                            .write(fileContent)
+                            .closeBlock("end");
+
+                    context.fileManifest().writeFile(fileName, writer.toString());
+                    return Collections.singletonList(relativeName);
+                } catch (IOException e) {
+                    throw new CodegenException(
+                            "Error reading rubySource file: " + rubyFileName,
+                            e);
+                }
+            };
+            return this;
+        }
+
+        /**
+         * Plugins may be implemented via classes with the `call(config)` method.
+         * Setting the pluginKlass will initialize the provided class with no arguments
+         * and add it to the list of client class plugins.
+         *
+         * @param pluginKlass - the fully qualified class name of the plugin to add
+         * @return this builder
+         */
+        public Builder pluginKlass(String pluginKlass) {
```

Comment:
Its not reserved, but I think we have generally used Klass for places where a String describes a class name, rather than  for example, a Java Class (ie, if this was `public Builder pluginClass(Class pluginClass)`). 

Open to changing that though.

Reply:
Yeah personally I think using klass only makes sense to avoid reserved naming, but pluginClass isn't reserved. If this were Builder klass(Class klass) it would make sense. It's up to you, I'd like to just be consistent.

Reply:
Updated to pluginClass.